Bencodex JSON Representation (#5)
planetarium/bencodex: JSON.md

Bencodex JSON Representation
===========================

## Introduction

> Bencodex JSON Representation is a standard way to represent a given Bencodex value to valid JSON text to be interchanged.

Reviewer: To make it clear that the transformation is processed at data model, not parsing:

```suggestion
Bencodex JSON Representation is a standard way to represent a given Bencodex value to a valid JSON value to be interchanged.
```

This specification defines a standard way to represent a Bencodex value to (a subset of) JSON and how to decode/encode them.

### Conventions Used in This Document

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in BCP 14[^RFC-2119] when, and only when, they appear in all capitals, as shown here.

## Prefixes

Bencodex JSON Representation defines a set of prefixes that **SHALL** be used for JSON[^JSON] string value to denote a corresponding Bencodex type. The defined prefixes are following:

> - `0x` denotes a hexadecimal byte-string. The rest of the string **SHOULD** be parsed as a byte-string encoded in hexadecimal.
>   - The casing of the value **MAY** be ignored.

Reviewer: IMHO the way to handle casing should be done on both side:

```suggestion
- `0x` denotes a hexadecimal byte string. The rest of the string **SHOULD** be parsed as a byte-string encoded in hexadecimal.
  - The hexadecimal byte string **SHOULD NOT** contain any upper letters.
  - The casing of the value **SHOULD** be ignored.
```

Author: I was thinking about some use-cases where the casings could be helpful such as [EIP-55](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-55.md) but as it is specified as `SHOULD` I agree with you.

> - `0x` denotes a hexadecimal byte-string. The rest of the string **SHOULD** be parsed as a byte-string encoded in hexadecimal.
>   - The casing of the value **MAY** be ignored.
> - `U+FEFF BYTE ORDER MARK` denotes a string. The rest of the string **MUST** be parsed as a string.
> - `b64:` denotes a base64-encoded byte-string. The rest of the string **MUST** be parsed as a byte-string encoded in base64 as defined in [RFC 4648](https://www.rfc-editor.org/rfc/rfc4648.html)[^RFC-4648].

Reviewer: It's trivial, but it would be better if we use consistent terms/spellings between the entire specification (see also *README.md*).

```suggestion
- `0x` denotes a hexadecimal byte string. The rest of the string **SHOULD** be parsed as a byte string encoded in hexadecimal.
  - The casing of the value **MAY** be ignored.
- `U+FEFF BYTE ORDER MARK` denotes a Unicode string. The rest of the string **MUST** be parsed as a Unicode string.
- `b64:` denotes a base64-encoded byte string. The rest of the string **MUST** be parsed as a byte string encoded in base64 as defined in [RFC 4648](https://www.rfc-editor.org/rfc/rfc4648.html)[^RFC-4648].
```

- In case no prefixes mentioned above are found, the value of the string **MUST** be parsed as an integer and having any non-digit character in the string **SHOULD** result in an error.

If the JSON string value contains escaped Unicode code points (denoted using `\u`) they **SHOULD** be interpreted before identifying a prefix.

## Encoding to JSON

When encoding a given value to JSON, the value **MUST** be encoded as specified in the following table.

| Bencodex type      | JSON type                 |
| ------------------ | ------------------------- |
| Null (`n`)         | Null (`null`)             |
| Boolean (`t`, `f`) | Boolean (`true`, `false`) |
| Byte string        | See [Prefixes](#prefixes) |
| Unicode string     | See [Prefixes](#prefixes) |

> | Integer (`i42e`)   | See [Prefixes](#prefixes) |
> | Lists (`l`-`e`)    | Array (`[]`)              |

Reviewer: ```suggestion
| Integer (`i`–`e`)   | See [Prefixes](#prefixes) |
| Lists (`l`–`e`)    | Array (`[]`)              |
```

"SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this
document are to be interpreted as described in BCP 14[^RFC-2119] when, and only
when, they appear in all capitals, as shown here.

Prefixes
--------

Bencodex JSON Representation defines a set of prefixes that **SHALL** be used
for JSON[^JSON] string value to denote a corresponding Bencodex type. The
defined prefixes are following:

- `0x` denotes a hexadecimal byte string. The rest of the string **SHOULD** be
parsed as a byte-string encoded in hexadecimal.
  - The hexadecimal byte string **SHOULD NOT** contain any upper letters.
  - The casing of the value **SHOULD** be ignored.
- `U+FEFF BYTE ORDER MARK` denotes a Unicode string. The rest of the string
**MUST** be parsed as a Unicode string.
- `b64:` denotes a base64-encoded byte string. The rest of the string **MUST**
be parsed as a byte string encoded in base64 as defined in
[RFC 4648](https://www.rfc-editor.org/rfc/rfc4648.html)[^RFC-4648].
- In case no prefixes mentioned above are found, the value of the string
**MUST** be parsed as an integer and having any non-digit character in the

> string **SHOULD** result in an error.

Reviewer: We should deal with negative integers as well.
